airflow/jobs/job.py

```diff
@@ -183,30 +184,32 @@ def heartbeat(
         previous_heartbeat = self.latest_heartbeat
 
         try:
-            # This will cause it to load from the db
-            self._merge_from(Job._fetch_from_db(self, session))
-            previous_heartbeat = self.latest_heartbeat
-
-            if self.state == JobState.RESTARTING:
-                self.kill()
-
-            # Figure out how long to sleep for
-            sleep_for = 0
-            if self.latest_heartbeat:
-                seconds_remaining = (
-                    self.heartrate - (timezone.utcnow() - self.latest_heartbeat).total_seconds()
-                )
-                sleep_for = max(0, seconds_remaining)
-            sleep(sleep_for)
-
-            job = Job._update_heartbeat(job=self, session=session)
-            self._merge_from(job)
-
-            # At this point, the DB has updated.
-            previous_heartbeat = self.latest_heartbeat
-
-            heartbeat_callback(session)
-            self.log.debug("[heartbeat]")
+            for attempt in run_with_db_retries(logger=self.log):
+                with attempt:
+                    # This will cause it to load from the db
+                    self._merge_from(Job._fetch_from_db(self, session))
```

Comment:
This is an example where we have problem with in-memory state and DB. You should not wrap both DB calls here with retry - because we can have some consistency problems here. you should retry fetch separately, update_heartbeat separatley I think.

Reply:
Yes that make sense to add retries directly on the method that make the db calls. I can look at adding that